README
PAR-Repository
==============

This module is intended for creation and maintenance of PAR repositories.
A PAR repository is collection of .par archives which contain Perl code
and associated libraries for use on specific platforms. In the most common
case, these archives differ from CPAN distributions in that they ship the
(possibly compiled) output of "make" in the "blib/" subdirectory of the
CPAN distribution's build directory.

You can access a PAR repository using the PAR::Repository::Client module
or the PAR module which provides syntactic sugar around the client.
PAR allows you to load libraries from repositories on demand.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires other modules and libraries. They are listed in
the Makefile.PL file with associated versions.

COPYRIGHT AND LICENCE

Copyright (C) 2006-2008 by Steffen Müller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.




Changes
0.20  Wed Aug 26 15:19 2009
    - Many temporary-file-locking fixes to make
      PAR::Repository work on FreeBSD (Slaven Rezic)
    - Fix to the test count for MD5 checksums (Slaven Rezic)
    - Documentation/POD fixes (Slaven Rezic)
    - Dependency fix for File::Temp. Require 0.19 (Slaven Rezic)

0.19  Thu Feb 19 16:48 2009
    - Promote to stable version.

0.18_01  Mon Feb  2 02:46 2009
    - Add ignore_provides switch to package injection.
    - Add static dependencies
    - Remove warning in Makefile.PL about symlink support
    - Fix bug in _update_info_version.
    - Add _create_dbm as an abstraction.
    - Slight fixes to 'parrepo' (verbosity)

0.17  Tue Jan 13 19:48 2009
    - Do a chmod(644) of the dbm checksums file after creation.

0.16_02  Thu Oct 23 20:51 2008
    - Occasionally, using defined($hash->{$foo}), we were inserting bogus
      records into the dbs.
    - Archive injection/removal test with symlink conversion in between
    - Slight refactoring of the test code.

0.16_01  Mon Oct 21 00:16 2008
    - Support for running without symlinks finally arrived!
    - fake_symlinks option to the constructor (default if
      symlinks not available)
    - convert_symlinks option to the constructor for converting
      existing repositories with symlinks to those without.
    - Fix to the ::DBM docs: the symlinks DB doesn't have paths.
    - Optimize the symlinks db space usage after deleting from it.
    - Expose the new options with parrepo (with a new command
      convert_symlinks, too)
    - Much improved test suite.
    - No longer use base if @ISA is ample.

0.16     Thu Sep  4 16:21 2008
    - Fix the documentation in ::DBM on the symlinks database.
    - Fix the name of the env. variable for disabling the
      no-symlinks warning on installation. (Cosimo Streppone)
    - Removed the PAR::Repository::ScanPAR module and use the
      PAR::Indexer module from CPAN instead. PAR::Indexer
      is essentially just PAR::Repository::ScanPAR, but usable
      by other distributions.
    - Fixed fallback to Digest::Perl::MD5
    - Finally add some meaningful tests.

0.15     Sun Aug 10 11:48 2008
    - Properly close scripts dbm on DESTROY.
    - Calculate DBM checkums.
    - Require DBM::Deep 1.0012: No more XS dependency.

0.14     Sat Nov  4 22:47 2006
    - Fixed a bug during injection.
    - META.yml now conforms to spec. 1.20.
    - Now including t/00podcover.t

0.13     Fri Sep 15 10:50 2006
    - Broke out PAR::Repository::Query into its own distribution
      so that PAR::Repository and PAR::Repository::Client can share
      the same querying interface.
    - Slight change to the interface!
      modules_dbm(), symlinks_dbm(), and scripts_dbm() now all return
      the name of the temporary file as second return value in order
      to make *_dbm() work consistently across PAR::Repository::Query
      uses.

0.12     Sun Aug 28 10:31 2006
    - Explicitly depending on PAR::Dist 0.18.
    - Added optional POD tests.
    - Fixed spelling error in query output of parrepo

0.11     Sat Aug 28 19:18 2006
    - Relaxed the "query dist" command syntax a bit.

0.10     Sat Aug 28 18:26 2006
    - Implemented querying.
      --> Added Query.pm
      --> Added "query script", "query module" and "query dist" to parrepo

0.03     Thu Aug 10 18:01 2006
    - Added scripts support.
    - Added support for upgrading repository versions.
    - Fixed a dangerous bug in the remove() method. (It didn't work
      correctly for symlinks before!)
    - Removing files from the repository by specifying individual bits
      of information now works.

0.02     Thu Aug 10 15:00 2006
    - Added new file "repository_info.yml" to the repositories.
    - Checks for compatible repository versions when opening a
      repository.
    - Corrected a couple of error messages.
    
0.01_02  Sat Aug 05 18:34 2006
    - Added documentation for the remove command in "parrepo".
    - Corrected documentation error.
    - Now including POD correctness tests.

0.01     Tue Jul 25 14:21 2006
    - original version



t/lib/RepoTest.pm
package RepoTest;
use strict;
use warnings;

use File::Temp;
use File::Spec;
use Cwd ();
require Test::More;

use vars qw/$StartDir $Debug/;

BEGIN {$StartDir = Cwd::cwd();}

my ($datadir, $parrepo_cmd, @test_dists, $tempdir);

my @repofiles = qw(
  dbm_checksums.txt
  modules_dists.dbm.zip
  repository_info.yml
  scripts_dists.dbm.zip
  symlinks.dbm.zip
);

setup();

sub setup {
  $datadir = File::Spec->rel2abs('data');
  Test::More::ok(-d $datadir, __PACKAGE__ . ': located data dir');

  $parrepo_cmd = File::Spec->rel2abs(
    File::Spec->catfile(
      File::Spec->updir(),
      'bin',
      'parrepo',
    )
  );
  Test::More::ok(-f $parrepo_cmd, __PACKAGE__ . ': located parrepo command');

  opendir my $dh, $datadir or die $!;
  while (defined(my $file = readdir($dh))) {
    my $full = File::Spec->catfile($datadir, $file);
    next if not -f $full;
    push @test_dists, $full if $full =~ /\.par$/i;
  }
  closedir $dh;
  Test::More::ok(scalar(@test_dists), __PACKAGE__ . ': located one or more test distributions');

}

sub RepoFiles { [@repofiles] }
sub DataDir { $datadir }
sub ParrepoCmd { $parrepo_cmd }
sub TestDists { [@test_dists] }

sub TempDir {
  return $tempdir if defined $tempdir;
  $tempdir = File::Temp::tempdir( CLEANUP => 1 );
  return $tempdir;
}

sub RunParrepo {
  my $class = shift;
  my @args = @_;
  my $cmd = RepoTest->ParrepoCmd;
  my $perl = $^X;
  my $inc = File::Spec->catdir($StartDir, File::Spec->updir(), 'blib', 'lib');
  my @full_cmd = (
    $perl,
    "-I$inc",
    $cmd,
    @args
  );

  warn __PACKAGE__ . ": Running '@full_cmd'" if $Debug;

  return system(@full_cmd);
}

sub TestRepoFilesExist {
  my $class = shift;
  my $path = shift;
  my $files = RepoTest->RepoFiles;

  foreach my $file (@$files) {
    my $full = File::Spec->catfile($path, $file);
    Test::More::ok(-f $full, __PACKAGE__ . ": parrepo contains file '$file'");
  }
}

sub CanOpenRepo {
  my $class = shift;
  require PAR::Repository;

  my $path = shift;
  my @args = @_;

  my $repo = PAR::Repository->new( path => $path, @args );
  Test::More::isa_ok($repo, 'PAR::Repository');
  Test::More::ok(-d $path, __PACKAGE__ . ': repo path exists after open');
  return $repo;
}

sub ConvertSymlinks {
  my $class = shift;
  my $path = shift;
  my $repo = $class->CanOpenRepo($path);
  
  $repo->_convert_symlinks();
  return();
}



END { chdir($StartDir); }

1;



t/lib/RepoTest/TestKit.pm
package RepoTest::TestKit;
use strict;
use warnings;

require Test::More;
require File::Spec;

####################
sub check_injection {
  my $class = shift;
  my $files = shift;
  die if not defined $files;
  $files = [$files] if not ref($files);
  my $repodir = shift || File::Spec->catdir(RepoTest->TempDir, 'repo');

  foreach my $file (@$files) {
    my ($dn, $dv, $arch, $pv) = PAR::Dist::parse_dist_name($file);
    Test::More::ok(
      -e File::Spec->catfile($repodir, $arch, $pv, $file),
      "par '$file' was injected"
    );
  }

  # test whether the stuff is in the repository now
  my $repo = RepoTest->CanOpenRepo($repodir);
  my $result = {@{$repo->query_module(regex => '^Test::Kit$')}};
  my $expect = {map {($_, '0.02')} @$files};
  Test::More::is_deeply(
    $result, $expect
  );
  #use Data::Dumper; warn Dumper $result;
  #use Data::Dumper; warn Dumper $expect;

  $result = {@{$repo->query_dist(regex => '^Test-Kit')}};
  $expect = {map {($_,
      {
        'Test::Kit' => '0.02',
        'Test::Kit::Result' => '0.02',
        'Test::Kit::Features' => '0.02',
      },
      )} @$files
  };
  Test::More::is_deeply(
    $result, $expect
  );

}

####################
sub check_symlinks {
  my $class = shift;
  my $symlinks = shift || {};
  my $repodir = shift || File::Spec->catdir(RepoTest->TempDir, 'repo');

  my $repo = RepoTest->CanOpenRepo($repodir);

  my ($dbm) = $repo->symlinks_dbm();
  my $copy = tied(%$dbm)->export(); # don't do this at home;
  Test::More::is_deeply(
    $copy,
    $symlinks
  );
}

####################
sub check_removal {
  my $class = shift;
  my $file = shift;
  die if not defined $file;
  my $repodir = shift || File::Spec->catdir(RepoTest->TempDir, 'repo');

  my ($dn, $dv, $arch, $pv) = PAR::Dist::parse_dist_name($file);

  my $repo = RepoTest->CanOpenRepo($repodir);
  Test::More::ok(
    !-f File::Spec->catfile($repodir, $arch, $pv, $file),
    'par was removed'
  );
  Test::More::is_deeply(
    $repo->query_module(regex => '^Test::Kit$'),
    [],
  );
  Test::More::is_deeply(
    $repo->query_dist(regex => '^Test-Kit'),
    []
  );
}

1;



Makefile.PL
use 5.006;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile(
    NAME              => 'PAR::Repository',
    VERSION_FROM      => 'lib/PAR/Repository.pm', # finds $VERSION
    LICENSE => 'perl',
    PREREQ_PM         => {
#'perl'      => '5.6',
        'DBM::Deep' => '1.0012',
        'File::Spec' => '0',
        'File::Path' => '0',
        'File::Temp' => '0',
        'File::Copy' => '0',
        'Archive::Zip' => '0',
        'PAR::Dist' => '0.44',
        'YAML::Syck' => '0.62',
        'version' => '0.50',
        'Pod::Text' => '0',
        'ExtUtils::Manifest' => '0',
        'PAR::Repository::Query' => '0.12',
        'PAR::Indexer' => '0.91',
        'Digest::MD5' => '0',
    }, # e.g., Module::Name => 1.1
    EXE_FILES => [
      'bin/parrepo',
    ],
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/PAR/Repository.pm', # retrieve abstract from module
       AUTHOR         => 'Steffen Müller <smueller@cpan.org>') : ()),
);


lib/PAR/Repository/Zip.pm
package PAR::Repository::Zip;

use 5.006;
use strict;
use warnings;

use Carp qw/croak/;
use File::Spec;
use File::Temp;
#use File::Path qw/rmtree/;
#use ExtUtils::Manifest;
#require ExtUtils::MM;
use Archive::Zip;

our $VERSION = '0.18';

=head1 NAME

PAR::Repository::Zip - ZIP file handling routines for PAR::Repository

=head1 SYNOPSIS

  use PAR::Repository;
  ...

=head1 DESCRIPTION

This module is for internal use only.
It contains code for dealing with ZIP files. .par archives
are ZIP files.

B<All methods here-in are considered private methods (note the
underscores in the names). Do not use outside of PAR::Repository.>
(Of course, you are free to copy the source code (see license).)

=head2 EXPORT

None.

=head1 METHODS

Following is a list of class and instance methods.
(Instance methods until otherwise mentioned.)

There is no C<PAR::Repository::Zip> object.
L<PAR::Repository> inherits from this class.

=cut

=head2 _unzip_dist_to_path

First argument should be path and file name of a .par
distribution. Second argument may be a path to unzip the
distribution to. (Defaults to current working directory.)

Unzips the distribution to the specified directory and returns
the directory name. Returns the empty list on failure.

=cut

sub _unzip_dist_to_path {
  my $self = shift;
  $self->verbose(2, "Entering _unzip_to_path()");
  my $dist = shift;
  my $path = shift || File::Spec->curdir;
  return unless -f $dist;

  my $zip = Archive::Zip->new;
  local %SIG;
  $SIG{__WARN__} = sub { print STDERR $_[0] unless $_[0] =~ /\bstat\b/ };
  return
    unless $zip->read($dist) == Archive::Zip::AZ_OK()
           and $zip->extractTree('', "$path/") == Archive::Zip::AZ_OK();
  return $path;
}

=head2 _unzip_dist_to_tmpdir

Creates a temporary directory and extracts a .par/zip archive into it.
First argument must be the archive file and (optional) second argument may
be a sub directory (of the temp dir) to extract into. This is mainly intended for
C<blib/> sub directories.

=cut

sub _unzip_dist_to_tmpdir {
  my $self = shift;
  $self->verbose(2, "Entering _unzip_dist_to_tmpdir()");
  my $dist   = File::Spec->rel2abs(shift);
  my $subdir = shift;
  my $tmpdir = File::Temp::mkdtemp(File::Spec->catdir(File::Spec->tmpdir, "parXXXXX")) or die $!;
  my $path = $tmpdir;
  $path = File::Spec->catdir($tmpdir, $subdir) if defined $subdir;
  $self->_unzip_dist_to_path($dist, $path);

  chdir $tmpdir;
  return ($dist, $tmpdir);
}

=head2 _zip_file

Callable as class or instance method.

Zips the file given as first argument to the file
given as second argument. If there is no second argument,
zips to "file1.zip" where "file1" was the first argument.

Returns the name of the zip file.

Optional third argument is the zip member name to use.

=cut

sub _zip_file {
  my $class = shift;
  my $file = shift;
  return unless -f $file;
  my $target = shift;
  my $member_name = shift;
  $member_name = $file if not defined $member_name;
  $target = $file.'.zip' if not defined $target;

  my $zip = Archive::Zip->new;
  my $member = $zip->addFile( $file, $member_name );
  $member->desiredCompressionLevel( Archive::Zip::COMPRESSION_LEVEL_BEST_COMPRESSION() );
  $zip->writeToFileNamed( $target ) == Archive::Zip::AZ_OK() or die $!;
  
  return $target;
}

=head2 _unzip_file

Unzips the file given as first argument to the file
given as second argument.
If a third argument is used, the zip member of that name
is extracted. If the zip member name is omitted, it is
set to the target file name.

Returns the name of the unzipped file.

=cut

sub _unzip_file {
  my $class = shift;
  my $file = shift;
  my $target = shift;
  my $member = shift;
  $member = $target if not defined $member;
  return unless -f $file;

  my $zip = Archive::Zip->new;
  local %SIG;
  $SIG{__WARN__} = sub { print STDERR $_[0] unless $_[0] =~ /\bstat\b/ };
      
  return unless $zip->read($file) == Archive::Zip::AZ_OK()
         and $zip->extractMember($member, $target) == Archive::Zip::AZ_OK();

  return $target;
}



1;
__END__

=head1 AUTHOR

Steffen Müller, E<lt>smueller@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2006-2009 by Steffen Müller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/PAR/Repository/DBM.pm
package PAR::Repository::DBM;

use 5.006;
use strict;
use warnings;

use Carp qw/croak/;
use File::Spec::Functions qw/catfile splitpath/;
use DBM::Deep;
use Fcntl qw/:flock/;
use File::Copy qw();

our $VERSION = '0.18';
use constant 'MODULES_DBM_FILE'      => 'modules_dists.dbm';
use constant 'SYMLINKS_DBM_FILE'     => 'symlinks.dbm';
use constant 'SCRIPTS_DBM_FILE'      => 'scripts_dists.dbm';
use constant 'DEPENDENCIES_DBM_FILE' => 'dependencies.dbm';
use constant 'DBM_CHECKSUMS_FILE'    => 'dbm_checksums.txt';

=head1 NAME

PAR::Repository::DBM - DBM tools for PAR::Repository

=head1 SYNOPSIS

  use PAR::Repository;

=head1 DESCRIPTION

This module is for internal use only.
It contains code for accessing the DBM files of a
PAR repository.

=head2 EXPORT

None.

=head2 GLOBALS

This package has a few constants:

C<MODULES_DBM_FILE>, C<SYMLINKS_DBM_FILE>, and C<SCRIPTS_DBM_FILE>,
C<DEPENDENCIES_DBM_FILE>, and C<DBM_CHECKSUMS_FILE>.
They are accessible
as functions via C<PAR::Repository::DBM::...>. They indicate
the file names of the DBM databases and the DBM checksums file.

=head1 DATABASE STRUCTURE

This section outlines the structure of the DBM::Deep
database files used by PAR::Repository.

If you need to care about this, you should be a
PAR::Repository developer.

=head2 MODULES-DISTS DBM

The DBM file is a hash at top level.

It associates namespaces (keys) with a number of file names and
versions. The values of the top level hash are hashes again. These
contain file names as keys and corresponding versions as values.

Example:

  {
    'Math::Symbolic::Derivative' => {
      'Math-Symbolic-0.502-x86_64-linux-gnu-thread-multi-5.8.7.par' => '0.502',
      'Math-Symbolic-0.200-x86_64-linux-gnu-thread-multi-5.8.6.par' => '0.200',
    },
  }

This example means that the C<Math::Symbolic::Derivative> module can
be found in the two listed distribution files in the repository
with the listed versions. Note that the distribution version needs not
be the same as the B<module> version. The module version is the one
separately indicated.

=head2 SYMLINKS DBM

The DBM file is a hash at top level.

It associates real files in the repository (keys) with a number of
symbolic links.
The values of the top level hash are arrays of distribution file names
which are symlinks.

Example: (with some extra linebreaks to keep the text width down)

  {
    'Math-Symbolic-0.502-x86_64-linux-gnu-thread-multi-5.8.7.par'
       => [
            'Math-Symbolic-0.502-any_arch-5.8.7.par',
            'Math-Symbolic-0.502-x86_64-linux-gnu-thread-multi-'
            .'any_version.par',
            'Math-Symbolic-0.502-any_arch-any_version.par'
          ],
  }

In the example, the first file is the real file and the paths/file names
in the value array are the names of the symbolic links.

=head2 SCRIPTS-DISTS DBM

This DBM file is a hash at top level. It associates script (executable)
names with distributions much like the C<modules_dists.dbm> file.

Example:

  {
    'parrepo' => {
      'PAR-Repository-0.03-x86_64--5.8.7.par'        => '0.02',
      'PAR-Repository-0.02-any_arch-any_version.par' => '0.01',
    },
  }

=head2 DEPENDENCIES DBM

This DBM file stores distribution names and associates them with
names of modules it depends on and their minimum versions. It does
not differentiate between the various types of dependencies that
can be found in a CPAN C<META.yml> file.

Example:

 {
   'Distname-0.03-x86_64-any_version.par' => {
      'Module::It::Depends::On' => '1.00',
   },
 }

=head1 METHODS

Following is a list of class and instance methods.
(Instance methods until otherwise mentioned.)

There is no C<PAR::Repository::DBM> object.
L<PAR::Repository> inherits from this class.

=cut

=head2 modules_dbm

Opens the modules_dists.dbm.zip file in the repository and
returns a tied hash reference to that file. Second return value is the
file name.

If the file does not exist, it returns the empty list.

You should know what you are doing when you use this
method.

=cut

sub modules_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering modules_dbm()');

  if (defined $self->{modules_dbm_hash}) {
    return $self->{modules_dbm_hash};
  }

  my $old_dir = Cwd::cwd();
  chdir($self->{path});
  my $file = PAR::Repository::DBM::MODULES_DBM_FILE().'.zip';
  chdir($old_dir), return() if not -f $file;

  my ($hash, $tempfile) = $self->_open_dbm($file);
  chdir($old_dir), return() if not defined $hash;

  $self->{modules_dbm_hash} = $hash;
  $self->{modules_dbm_temp_file} = $tempfile;

  chdir($old_dir);

  return ($hash, $tempfile);
}


=head2 symlinks_dbm

Opens the symlinks.dbm.zip file in the repository and
returns a tied hash reference to that file. Second
return value is the file name.

If the file does not exist, it returns the empty list.

You should know what you are doing when you use this
method.

=cut

sub symlinks_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering symlinks_dbm()');

  if (defined $self->{symlinks_dbm_hash}) {
    return $self->{symlinks_dbm_hash};
  }

  my $old_dir = Cwd::cwd();
  chdir($self->{path});

  my $file = PAR::Repository::DBM::SYMLINKS_DBM_FILE().'.zip';

  chdir($old_dir), return() if not -f $file;

  my ($hash, $tempfile) = $self->_open_dbm($file);
  chdir($old_dir), return() if not defined $hash;

  $self->{symlinks_dbm_hash} = $hash;
  $self->{symlinks_dbm_temp_file} = $tempfile;

  chdir($old_dir);

  return($hash, $tempfile);
}

=head2 scripts_dbm

Opens the scripts_dists.dbm.zip file in the repository and
returns a tied hash reference to that file. Second return value is
the file name.

If the file does not exist, it returns the empty list.

You should know what you are doing when you use this
method.

=cut

sub scripts_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering scripts_dbm()');

  if (defined $self->{scripts_dbm_hash}) {
    return $self->{scripts_dbm_hash};
  }

  my $old_dir = Cwd::cwd();
  chdir($self->{path});
  my $file = PAR::Repository::DBM::SCRIPTS_DBM_FILE().'.zip';
  chdir($old_dir), return() if not -f $file;

  my ($hash, $tempfile) = $self->_open_dbm($file);
  chdir($old_dir), return() if not defined $hash;

  $self->{scripts_dbm_hash} = $hash;
  $self->{scripts_dbm_temp_file} = $tempfile;

  chdir($old_dir);

  return($hash, $tempfile);
}

=head2 dependencies_dbm

Opens the dependencies.dbm.zip file in the repository and
returns a tied hash reference to that file. Second return value is the
file name.

If the file does not exist, it returns the empty list.

You should know what you are doing when you use this
method.

=cut

sub dependencies_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering dependencies_dbm()');

  if (defined $self->{dependencies_dbm_hash}) {
    return $self->{dependencies_dbm_hash};
  }

  my $old_dir = Cwd::cwd();
  chdir($self->{path});
  my $file = PAR::Repository::DBM::DEPENDENCIES_DBM_FILE().'.zip';
  chdir($old_dir), return() if not -f $file;

  my ($hash, $tempfile) = $self->_open_dbm($file);
  chdir($old_dir), return() if not defined $hash;

  $self->{dependencies_dbm_hash} = $hash;
  $self->{dependencies_dbm_temp_file} = $tempfile;

  chdir($old_dir);

  return ($hash, $tempfile);
}


=head2 close_modules_dbm

Closes the C<modules_dists.dbm> file committing any
changes and then zips it back into
C<modules_dists.dbm.zip>.

This is called when the object is destroyed.

=cut

sub close_modules_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering close_modules_dbm()');
  my $hash = $self->{modules_dbm_hash};
  return if not defined $hash;

  my $obj = tied($hash);
  $self->{modules_dbm_hash} = undef;
  undef $hash;
  undef $obj;

  $self->_zip_file(
      $self->{modules_dbm_temp_file},
      catfile($self->{path}, PAR::Repository::DBM::MODULES_DBM_FILE().'.zip'),
      PAR::Repository::DBM::MODULES_DBM_FILE(),
  );

  unlink $self->{modules_dbm_temp_file};
  $self->{modules_dbm_temp_file} = undef;

  return 1;
}


=head2 close_symlinks_dbm

The same as C<close_modules_dbm()> but for the
file C<symlinks.dbm.zip>.

Also called on object destruction.

=cut

sub close_symlinks_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering close_symlinks_dbm()');
  my $hash = $self->{symlinks_dbm_hash};
  return if not defined $hash;

  my $obj = tied($hash);
  $self->{symlinks_dbm_hash} = undef;
  undef $hash;
  undef $obj;

  $self->_zip_file(
      $self->{symlinks_dbm_temp_file},
      catfile($self->{path}, PAR::Repository::DBM::SYMLINKS_DBM_FILE().'.zip'),
      PAR::Repository::DBM::SYMLINKS_DBM_FILE(),
  );

  unlink $self->{symlinks_dbm_temp_file};
  $self->{symlinks_dbm_temp_file} = undef;

  return 1;
}

=head2 close_scripts_dbm

Closes the C<scripts_dists.dbm> file committing any
changes and then zips it back into
C<scripts_dists.dbm.zip>.

This is called when the object is destroyed.

=cut

sub close_scripts_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering close_scripts_dbm()');
  my $hash = $self->{scripts_dbm_hash};
  return if not defined $hash;

  my $obj = tied($hash);
  $self->{scripts_dbm_hash} = undef;
  undef $hash;
  undef $obj;

  $self->_zip_file(
      $self->{scripts_dbm_temp_file},
      catfile($self->{path}, PAR::Repository::DBM::SCRIPTS_DBM_FILE().'.zip'),
      PAR::Repository::DBM::SCRIPTS_DBM_FILE()
  );

  unlink $self->{scripts_dbm_temp_file};
  $self->{scripts_dbm_temp_file} = undef;

  return 1;
}


=head2 close_dependencies_dbm

Closes the C<dependencies.dbm> file committing any
changes and then zips it back into
C<dependencies.dbm.zip>.

This is called when the object is destroyed.

=cut

sub close_dependencies_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering close_dependencies_dbm()');
  my $hash = $self->{dependencies_dbm_hash};
  return if not defined $hash;

  my $obj = tied($hash);
  $self->{dependencies_dbm_hash} = undef;
  undef $hash;
  undef $obj;

  $self->_zip_file(
      $self->{dependencies_dbm_temp_file},
      catfile($self->{path}, PAR::Repository::DBM::DEPENDENCIES_DBM_FILE().'.zip'),
      PAR::Repository::DBM::DEPENDENCIES_DBM_FILE(),
  );

  unlink $self->{dependencies_dbm_temp_file};
  $self->{dependencies_dbm_temp_file} = undef;

  return 1;
}


=head2 update_dbm_checksums

Updates the DBM checksums file C<dbm_checksums.txt> with the
checksums of the currently existing zipped DBM files.

This is called when the L<PAR::Repository> object is destroyed.

=cut

sub update_dbm_checksums {
  my $self = shift;
  $self->verbose(2, 'Entering update_dbm_checksums()');

  # find a working base64 MD5 implementation
  my $md5_function;
  eval { require Digest::MD5; $md5_function = \&Digest::MD5::md5_base64; };
  eval { require Digest::Perl::MD5;  $md5_function = \&Digest::Perl::MD5::md5_base64; } if $@;
  if ($@) {
    die "Could load neither Digest::MD5 nor Digest::Perl::MD5. Please upgrade your perl or install either of those modules.";
  }
  
  # Prepare temporary copy of the checkums file
  my ($tempfh, $tempfile) = File::Temp::tempfile(
      'temporary_dbm_checksum_XXXXX',
      UNLINK => 0,
      DIR    => File::Spec->tmpdir(),
  );
  print $tempfh <<'HERE';
# This checksums file has the format
# FILENAME BASE64_MD5_HASH
# where the file name and the MD5 hash are separated
# by a TAB character, not arbitrary whitespace!
HERE

  # calculate hashes and write them to the temp file
  foreach my $dbmfile (
      PAR::Repository::DBM::MODULES_DBM_FILE(),
      PAR::Repository::DBM::SCRIPTS_DBM_FILE(),
      PAR::Repository::DBM::SYMLINKS_DBM_FILE(),
      PAR::Repository::DBM::DEPENDENCIES_DBM_FILE(),
    ) {
    my $filepath = catfile($self->{path}, $dbmfile.'.zip');
    open my $fh, '<', $filepath or die "Could not open DBM file '$filepath' for reading: $!";
    flock $fh, LOCK_SH;
    local $/ = undef;
    my $hash = $md5_function->(<$fh>);
    close $fh;
    print $tempfh "$dbmfile.zip\t$hash\n";
  } # end foreach dbm files

  close $tempfh;
  # move temp file to destination
  my $target_file = catfile($self->{path}, PAR::Repository::DBM::DBM_CHECKSUMS_FILE());
  File::Copy::move($tempfile, $target_file)
    or die "Could not move checksums file '$tempfile' to '$target_file': $!";
  # FIXME, could this be done more user friendly? But somehow, the file ended up being 600 by default...
  chmod(0644, $target_file);

  return 1;
}


=head2 _open_dbm

Opens the zipped dbm file given as first argument.

This is B<only for internal use>.

=cut

sub _open_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering _open_dbm()');
  my $file = shift;
  my ($tempfh, $tempfile) = File::Temp::tempfile(
      'temporary_dbm_XXXXX',
      UNLINK => 0,
      DIR    => File::Spec->tmpdir(),
  );
  my ($v, $p, $f) = splitpath($file);
  $f =~ s/\.zip$//;
  $self->_unzip_file($file, $tempfile, $f) or return undef;
  my %hash;
  my $obj = tie %hash, "DBM::Deep", {
    file    => $tempfile,
    locking => 1,
  }; 

  return (\%hash, $tempfile);
}

=head2 _create_dbm

Creates a zipped dbm file given as first argument.

This is B<only for internal use>.

=cut

sub _create_dbm {
  my $self = shift;
  $self->verbose(2, 'Entering _create_dbm()');
  my $file = shift;
  $file .= '.zip' unless $file =~ /\.zip$/i;

  my ($tempfh, $tempfile) = File::Temp::tempfile(
      'temporary_dbm_XXXXX',
      UNLINK => 0,
      DIR    => File::Spec->tmpdir(),
  );
  {
    my %hash;
    my $obj = tie %hash, "DBM::Deep", {
      file    => $tempfile,
      locking => 1,
    }; 
  }

  my ($v, $p, $f) = splitpath($file);
  $f =~ s/\.zip$//i;
  $self->_zip_file($tempfile, $file, $f) or unlink($tempfile), return();

  unlink($tempfile);

  return 1;
}

1;
__END__

=head1 AUTHOR

Steffen Müller, E<lt>smueller@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2006-2009 by Steffen Müller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.

=cut
